### PR TITLE
Fixing issue: QOS not set for incoming messages from $iothub/twin/PAT…

### PR DESCRIPTION
**Pull request by dzakub**
…CH/properties/desired/ topic #1746

 

# Checklist
- [x] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-java/blob/main/.github/CONTRIBUTING.md).
- [x] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- I submitted this PR against the correct branch: 
- [x] This pull-request is submitted against the `main` branch. 

# Reference/Link to the issue solved with this PR (if any)
https://github.com/Azure/azure-iot-sdk-java/issues/1746

# Description of the problem
Problem is described in the issue.

# Description of the solution
Added code to copy QOS from original message.

### CODE REVIEW

**Comment by dzakub** [comment]
@microsoft-github-policy-service agree

File: iothub/device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttTwinTest.java [hunk 1]
```diff
@@ -904,6 +904,33 @@ public void receiveDoesNotSetVersionForDesiredPropNotifRespIfNotFound() throws T
         }
     }
 
+    @Test
+    public void receiveSetQosForDesiredProp() throws TransportException
+    {
+        final byte[] actualPayload = "NotificationResponseDataContainingDesiredPropertiesDocument".getBytes(StandardCharsets.UTF_8);
+        final String expectedTopic = "$iothub/twin/PATCH/properties/desired/";
+        IotHubTransportMessage receivedMessage = null;
+        int expectedQualityOfService = 2;
```
Comment (timtay-microsoft) [inline iothub/device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttTwinTest.java:913]: Small nitpick to make this unit test closer to what an actual received twin message would look like.

```suggestion
        int expectedQualityOfService = 0;
```

**Comment by timtay-microsoft** [comment]
This change seems reasonable, but these messages always have a QoS of 0, so I'm not sure what change in behavior you'll see from this.

**Review by timtay-microsoft — APPROVED** [review]

File: iothub/device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttTwinTest.java (hunk 1 above)
Comment (dzakub) [reply iothub/device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttTwinTest.java:913]: In other case I would agree - I'm all for making test data as close to actual production data as possible. But doing so, I would made this test passing regarding of production code change - QOS is held as `int` which during object construction is set to default value: 0.

First I created failing test, and the fix production making test green - this way I knew that test is actually checking case I wanted to cover.

**Comment by dzakub** [comment]
> This change seems reasonable, but these messages always have a QoS of 0, so I'm not sure what change in behavior you'll see from this. Are you expecting QoS other than 0 here?

I didn't know that IOT hub sends always massages with QOS 0 here. If there will be no change in behavior.
I discovered this 'issue' whlie stubbing iot hub for integration tests that were written to document behavior of 1.x clients and to ease upgrading to 2.x. 
Tests were green for 1.x, but started failing after upgrade to 2.x. After investigation I discovered mentioned problem (my fake iot hub was sending messages with QOS 1)

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttTwin.java [hunk 2]
```diff
@@ -342,6 +342,7 @@ else if (topic.length() > PATCH.length() && topic.startsWith(PATCH))
                                 {
                                     message = new IotHubTransportMessage(data, MessageType.DEVICE_TWIN);
                                     message.setDeviceOperationType(DeviceOperations.DEVICE_OPERATION_TWIN_SUBSCRIBE_DESIRED_PROPERTIES_RESPONSE);
+                                    message.setQualityOfService(messagePair.getValue().getQos());
```
[the remark this answers is not in the document]
Comment (dzakub) [reply iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttTwin.java:345]: Sure, thank you.

**Pull request merged**